rust/src/transformations/make_stable_expr/expr_fill_null/mod.rs

```diff
@@ -0,0 +1,153 @@
+
+/// Make a Transformation that returns a `fill_null(constant)` expression
+///
+/// # Arguments
+/// * `input_domain` - Expr domain
+/// * `input_metric` - The metric under which neighboring LazyFrames are compared
+/// * `expr` - The clipping expression
+pub fn make_expr_fill_null<M: OuterMetric>(
+    input_domain: ExprDomain,
+    input_metric: M,
+    expr: Expr,
+) -> Fallible<Transformation<ExprDomain, ExprDomain, M, M>>
+where
+    M::InnerMetric: DatasetMetric,
+    M::Distance: Clone,
+    (ExprDomain, M): MetricSpace,
+    Expr: StableExpr<M, M>,
+{
+    let Some((input, constant)) = match_fill_null(&expr)? else {
+        return fallible!(MakeTransformation, "expected fill_null expression");
+    };
+
+    let t_prior = input
+        .clone()
+        .make_stable(input_domain.clone(), input_metric.clone())?;
+    let (middle_domain, middle_metric) = t_prior.output_space();
+
+    let mut output_domain = middle_domain.clone();
+    let series_domain = output_domain.active_series_mut()?;
+
+    use DataType::*;
+    let constant = match &series_domain.field.dtype {
+        UInt32 => reconcile_domain::<u32>(constant, series_domain),
+        UInt64 => reconcile_domain::<u64>(constant, series_domain),
+        Int8 => reconcile_domain::<i8>(constant, series_domain),
+        Int16 => reconcile_domain::<i16>(constant, series_domain),
+        Int32 => reconcile_domain::<i32>(constant, series_domain),
+        Int64 => reconcile_domain::<i64>(constant, series_domain),
+        Float32 => reconcile_domain::<f32>(constant, series_domain),
+        Float64 => reconcile_domain::<f64>(constant, series_domain),
+        Boolean => reconcile_domain::<bool>(constant, series_domain),
+        String => reconcile_domain::<str>(constant, series_domain),
+        UInt8 | UInt16 => {
+            fallible!(
+                MakeTransformation,
+                "u8 and 16 are not supported, please use u32 or u64 instead"
+            )
+        }
+        dtype => fallible!(MakeTransformation, "unsupported dtype: {}", dtype),
+    }?;
+
+    t_prior
+        >> Transformation::new(
+            middle_domain.clone(),
+            output_domain,
+            Function::then_expr(move |expr| expr.fill_null(constant.clone())),
+            middle_metric.clone(),
+            middle_metric,
+            StabilityMap::new(Clone::clone),
+        )?
+}
+
+pub fn match_fill_null(expr: &Expr) -> Fallible<Option<(&Expr, &LiteralValue)>> {
+    let Expr::Function {
+        input,
+        function: FunctionExpr::FillNull { .. },
+        ..
+    } = expr
+    else {
+        return Ok(None);
+    };
+
+    let Ok([input, constant]) = <&[_; 2]>::try_from(input.as_slice()) else {
+        return fallible!(MakeTransformation, "fill_null expects 2 arguments");
+    };
+
+    let Expr::Literal(constant) = constant else {
+        return fallible!(
+            MakeTransformation,
+            "fill_null expects a literal as second argument"
+        );
+    };
```

Comment:
I'm expecting `fill_null` to be parallel to [`fill_null` in polars](https://docs.rs/polars/latest/polars/series/struct.Series.html#method.fill_null), but that talks about different fill "strategies", and points to [`fill_null_with_values`](https://docs.rs/polars/latest/polars/chunked_array/ops/trait.ChunkFillNullValue.html#tymethod.fill_null_with_values), which seems more like what we're doing here?

Reply:
There's a myriad of different APIs for DataFrame, LazyFrame, Series, ChunkedArray and Expr, but we're just supporting Expr for now.

Reply:
In the Expr::fill_null API, the value that you impute with is itself an Expr. I explicitly enforce that the expr is only a literal. This could possibly be relaxed!